### ISSUE TITLE
Master::async_client_init_ fails in TSAN on Shutdown

### ISSUE THREAD

**Issue by nspiegelberg** [yugabyte/yugabyte-db#5992]
I'm getting TSAN test failures in my code to support TLS connections with CDC on the CatalogManager (https://phabricator.dev.yugabyte.com/D9526).  While the failure is in TwoDCTest.ApplyOperationsRandomFailures, but the problem seems to be related to shutting down the YBClient connection.  With my change, YBClientBuilder::AddMasterAddressSource() is exercised for the first time (from https://phabricator.dev.yugabyte.com/D8323).  It seems to be unused code in Catalog Manager before.  

It is causing the following failures on TSAN builds:

```
F20201001 21:04:50 ../../src/yb/util/thread_restrictions.cc:91]  Check failed: LoadTLS()->wait_allowed Waiting is not allowed to be  used on this thread to prevent server-wide latency aberrations and  deadlocks. Thread 6285 (name: "Master_reactor", category:  "Master_reactor")
--
6439 | @     0x7f4532dd6c88  yb::LogFatalHandlerSink::send(int,  char const*, char const*, int, tm const*, char const*, unsigned long) (src/yb/util/logging.cc:474)
6440 | @     0x7f45321ec0dd
6441 | @     0x7f45321e3095
6442 | @     0x7f45321e3e5c
6443 | @     0x7f45321e8afa
6444 | @     0x7f4532ea44cd  yb::ThreadRestrictions::AssertWaitAllowed() (src/yb/util/thread_restrictions.cc:91)
6445 | @     0x7f453d3730aa  yb::consensus::ReplicaState::LockForRead() const (src/yb/consensus/replica_state.cc:137)
6446 | @     0x7f453d358e39   yb::consensus::RaftConsensus::ConsensusState(yb::consensus::ConsensusConfigType,  yb::consensus::LeaderLeaseStatus*) const (src/yb/consensus/raft_consensus.cc:2802)
6447 | @     0x7f4540d7179b  yb::master::CatalogManager::GetCurrentConfig(yb::consensus::ConsensusStatePB*) const (src/yb/master/catalog_manager.cc:7515)
6448 | @     0x7f4540e40f6e  operator() (src/yb/master/master.cc:184)
6449 | @     0x7f4540e40ee0   decltype(std::forward<yb::master::Master::Init()::$_0&>(fp)())   std::__invoke<yb::master::Master::Init()::$_0&>(yb::master::Master::Init()::$_0&)   (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/type_traits:4345)
6450 | @     0x7f4540e40ee0  std::vector<string,  std::allocator<string> >  std::__invoke_void_return_wrapper<std::vector<string,  std::allocator<string> >  >::__call<yb::master::Master::Init()::$_0&>(yb::master::Master::Init()::$_0&)   (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/__functional_base:318)
6451 | @     0x7f4540e40e24   std::__function::__func<yb::master::Master::Init()::$_0,  std::allocator<yb::master::Master::Init()::$_0>,  std::vector<string, std::allocator<string> >  ()>::operator()()  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/functional:1562)
6452 | @     0x7f453aa1958b   std::function<std::vector<string, std::allocator<string>  > ()>::operator()() const  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/functional:1913)
6453 | @     0x7f453a9d9cda  yb::client::YBClient::Data::ReinitializeMasterAddresses() (src/yb/client/client-internal.cc:1932)
6454 | @     0x7f453a9d8ff2   yb::client::YBClient::Data::SetMasterServerProxyAsync(std::chrono::time_point<yb::CoarseMonoClock,  std::chrono::duration<long long, std::ratio<1l, 1000000000l>  > >, bool, bool, yb::Callback<void (yb::Status const&)>  const&) (src/yb/client/client-internal.cc:1764)
6455 | @     0x7f453aa4447b  yb::client::internal::LookupRpc::ResetMasterLeaderAndRetry() (src/yb/client/meta_cache.cc:685)
6456 | @     0x7f453aa637e0  void  yb::client::internal::LookupRpc::DoFinished<yb::master::GetTabletLocationsResponsePB>(yb::Status  const&, yb::master::GetTabletLocationsResponsePB const&, string  const*) (src/yb/client/meta_cache.cc:752)
6457 | @     0x7f453aa62b49  yb::client::internal::LookupByIdRpc::Finished(yb::Status const&) (src/yb/client/meta_cache.cc:1123)
6458 | @     0x7f453aa6429b   decltype(*(std::forward<yb::client::internal::LookupByIdRpc*&>(fp0)).*fp(std::forward<yb::Status::OK&>(fp1)))  std::__invoke<void  (yb::client::internal::LookupByIdRpc::*&)(yb::Status const&),  yb::client::internal::LookupByIdRpc*&, yb::Status::OK&,  void>(void (yb::client::internal::LookupByIdRpc::*&)(yb::Status  const&), yb::client::internal::LookupByIdRpc*&,  yb::Status::OK&)  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/type_traits:4286)
6459 | @     0x7f453aa6429b  std::__bind_return<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  tuple<yb::client::internal::LookupByIdRpc*, yb::Status::OK>,  tuple<>, __is_valid_bind_return<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  tuple<yb::client::internal::LookupByIdRpc*, yb::Status::OK>,  tuple<> >::value>::type std::__apply_functor<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  tuple<yb::client::internal::LookupByIdRpc*, yb::Status::OK>, 0ul,  1ul, tuple<> >(void  (yb::client::internal::LookupByIdRpc::*&)(yb::Status const&),  tuple<yb::client::internal::LookupByIdRpc*, yb::Status::OK>&,  std::__tuple_indices<0ul, 1ul>, tuple<>&&)  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/functional:2216)
6460 | @     0x7f453aa6429b  std::__bind_return<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  tuple<yb::client::internal::LookupByIdRpc*, yb::Status::OK>,  tuple<>, __is_valid_bind_return<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  tuple<yb::client::internal::LookupByIdRpc*, yb::Status::OK>,  tuple<> >::value>::type std::__bind<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  yb::client::internal::LookupByIdRpc*,  yb::Status::OK>::operator()<>()  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/functional:2249)
6461 | @     0x7f453aa6429b   decltype(std::forward<std::__bind<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  yb::client::internal::LookupByIdRpc*, yb::Status::OK>&>(fp)())  std::__invoke<std::__bind<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  yb::client::internal::LookupByIdRpc*,  yb::Status::OK>&>(std::__bind<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  yb::client::internal::LookupByIdRpc*, yb::Status::OK>&)  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/type_traits:4345)
6462 | @     0x7f453aa6429b  void  std::__invoke_void_return_wrapper<void>::__call<std::__bind<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  yb::client::internal::LookupByIdRpc*,  yb::Status::OK>&>(std::__bind<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  yb::client::internal::LookupByIdRpc*, yb::Status::OK>&)  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/__functional_base:349)
6463 | @     0x7f453aa6418d   std::__function::__func<std::__bind<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  yb::client::internal::LookupByIdRpc*, yb::Status::OK>,  std::allocator<std::__bind<void  (yb::client::internal::LookupByIdRpc::*)(yb::Status const&),  yb::client::internal::LookupByIdRpc*, yb::Status::OK> >, void  ()>::operator()()  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/functional:1562)
6464 | @     0x7f453e95c6d4  std::function<void  ()>::operator()() const  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/functional:1913)
6465 | @     0x7f4533c7cf8a  yb::rpc::OutboundCall::InvokeCallbackSync() (src/yb/rpc/outbound_call.cc:383)
6466 | @     0x7f4533c7d1f1  yb::rpc::InvokeCallbackTask::Done(yb::Status const&) (src/yb/rpc/outbound_call.cc:173)
6467 | @     0x7f4533d0d1c2  yb::rpc::ThreadPool::Impl::Enqueue(yb::rpc::ThreadPoolTask*) (src/yb/rpc/thread_pool.cc:183)
6468 | @     0x7f4533d096b8  yb::rpc::ThreadPool::Enqueue(yb::rpc::ThreadPoolTask*) (src/yb/rpc/thread_pool.cc:290)
6469 | @     0x7f4533c7fa04  yb::rpc::OutboundCall::InvokeCallback() (src/yb/rpc/outbound_call.cc:368)
6470 | @     0x7f4533c7ee0f   yb::rpc::OutboundCall::SetFailed(yb::Status const&,  unique_ptr<yb::rpc::ErrorStatusPB,  std::default_delete<yb::rpc::ErrorStatusPB> >) (src/yb/rpc/outbound_call.cc:495)
6471 | @     0x7f4533c7fd51  yb::rpc::OutboundCall::SetResponse(yb::rpc::CallResponse&&) (src/yb/rpc/outbound_call.cc:440)
6472 | @     0x7f4533c4e230  yb::rpc::Connection::HandleCallResponse(yb::rpc::CallData*) (src/yb/rpc/connection.cc:325)
6473 | @     0x7f4533d13841   yb::rpc::YBOutboundConnectionContext::HandleCall(shared_ptr<yb::rpc::Connection>  const&, yb::rpc::CallData*) (src/yb/rpc/yb_rpc.cc:540)
6474 | @     0x7f4533d13899  non-virtual thunk to  yb::rpc::YBOutboundConnectionContext::HandleCall(shared_ptr<yb::rpc::Connection>  const&, yb::rpc::CallData*) (src/yb/rpc/yb_rpc.cc:0)
6475 | @     0x7f4533c45058   yb::rpc::BinaryCallParser::Parse(shared_ptr<yb::rpc::Connection>  const&, boost::container::small_vector<iovec, 4ul,  boost::container::new_allocator<iovec> > const&,  yb::StronglyTypedBool<yb::rpc::ReadBufferFull_Tag>,  shared_ptr<yb::MemTracker> const*) (src/yb/rpc/binary_call_parser.cc:154)
6476 | @     0x7f4533d13d64   yb::rpc::YBOutboundConnectionContext::ProcessCalls(shared_ptr<yb::rpc::Connection>  const&, boost::container::small_vector<iovec, 4ul,  boost::container::new_allocator<iovec> > const&,  yb::StronglyTypedBool<yb::rpc::ReadBufferFull_Tag>) (src/yb/rpc/yb_rpc.cc:561)
6477 | @     0x7f4533c4dd76   yb::rpc::Connection::ProcessReceived(boost::container::small_vector<iovec,  4ul, boost::container::new_allocator<iovec> > const&,  yb::StronglyTypedBool<yb::rpc::ReadBufferFull_Tag>) (src/yb/rpc/connection.cc:287)
6478 | @     0x7f4533d0560d  yb::rpc::TcpStream::TryProcessReceived() (src/yb/rpc/tcp_stream.cc:389)
6479 | @     0x7f4533d0497d  yb::rpc::TcpStream::ReadHandler() (src/yb/rpc/tcp_stream.cc:318)
6480 | @     0x7f4533d0457b  yb::rpc::TcpStream::Handler(ev::io&, int) (src/yb/rpc/tcp_stream.cc:260)
6481 | @     0x7f4533d07e1a  void ev::base<ev_io,  ev::io>::method_thunk<yb::rpc::TcpStream,  &(yb::rpc::TcpStream::Handler(ev::io&, int))>(ev_loop*,  ev_io*, int)  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/common/include/ev++.h:479)
6482 | @     0x7f4530c6db6a
6483 | @     0x7f4530c71c79
6484 | @     0x7f4533c3c898  ev::loop_ref::run(int)  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/common/include/ev++.h:211)
6485 | @     0x7f4533ca3d3c  yb::rpc::Reactor::RunThread() (src/yb/rpc/reactor.cc:482)
6486 | @     0x7f4533cb0253   decltype(*(std::forward<yb::rpc::Reactor*&>(fp0)).*fp())  std::__invoke<void (yb::rpc::Reactor::*&)(),  yb::rpc::Reactor*&, void>(void (yb::rpc::Reactor::*&)(),  yb::rpc::Reactor*&)  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/type_traits:4286)
6487 | @     0x7f4533cb0253  std::__bind_return<void  (yb::rpc::Reactor::*)(), tuple<yb::rpc::Reactor*>, tuple<>,  __is_valid_bind_return<void (yb::rpc::Reactor::*)(),  tuple<yb::rpc::Reactor*>, tuple<> >::value>::type  std::__apply_functor<void (yb::rpc::Reactor::*)(),  tuple<yb::rpc::Reactor*>, 0ul, tuple<> >(void  (yb::rpc::Reactor::*&)(), tuple<yb::rpc::Reactor*>&,  std::__tuple_indices<0ul>, tuple<>&&)  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/functional:2216)
6488 | @     0x7f4533cb0253  std::__bind_return<void  (yb::rpc::Reactor::*)(), tuple<yb::rpc::Reactor*>, tuple<>,  __is_valid_bind_return<void (yb::rpc::Reactor::*)(),  tuple<yb::rpc::Reactor*>, tuple<> >::value>::type  std::__bind<void (yb::rpc::Reactor::* const&)(),  yb::rpc::Reactor* const&>::operator()<>()  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/functional:2249)
6489 | @     0x7f4533cb0253   decltype(std::forward<std::__bind<void (yb::rpc::Reactor::*  const&)(), yb::rpc::Reactor* const&>&>(fp)())  std::__invoke<std::__bind<void (yb::rpc::Reactor::* const&)(),  yb::rpc::Reactor* const&>&>(std::__bind<void  (yb::rpc::Reactor::* const&)(), yb::rpc::Reactor*  const&>&)  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/type_traits:4345)
6490 | @     0x7f4533cb0253  void  std::__invoke_void_return_wrapper<void>::__call<std::__bind<void  (yb::rpc::Reactor::* const&)(), yb::rpc::Reactor*  const&>&>(std::__bind<void (yb::rpc::Reactor::*  const&)(), yb::rpc::Reactor* const&>&)  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/__functional_base:349)
6491 | @     0x7f4533cb015d   std::__function::__func<std::__bind<void (yb::rpc::Reactor::*  const&)(), yb::rpc::Reactor* const&>,  std::allocator<std::__bind<void (yb::rpc::Reactor::*  const&)(), yb::rpc::Reactor* const&> >, void  ()>::operator()()  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/functional:1562)
6492 | @     0x7f453e95c6d4  std::function<void  ()>::operator()() const  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/installed/tsan/libcxx/include/c++/v1/functional:1913)
6493 | @     0x7f4532e99ec9  yb::Thread::SuperviseThread(void*) (src/yb/util/thread.cc:760)
6494 | @           0x472c2d  __tsan_thread_start_func  (/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20200829090443-f431681041-centos/src/llvm-7.1.0.src/projects/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:955)
6495 | @     0x7f452e5e9693  start_thread (/tmp/glibc-20181130-26094-cs1x60/glibc-2.23/nptl/pthread_create.c:333)
6496 | @     0x7f452e11441c  (unknown) (sysdeps/unix/sysv/linux/x86_64/clone.S:109)
6497 | @ 0xffffffffffffffff
```

**Comment by spolitov** [comment]
@nspiegelberg `AddMasterAddressSource` is invoked from `Master::Init()`.
Why do you think that it was not exercised?

**Comment by nspiegelberg** [comment]
@spolitov : `git grep async_client_init_`.  While this code is used on the TServer in multiple locations, it is not actually invoked on the Master process anywhere.

**Comment by spolitov** [comment]
@nspiegelberg Could you please clarify this moment? Do you think that `Master::Init()` is not invoked in the Master process?

**Comment by nspiegelberg** [comment]
@spolitov : From my understanding, the `AddMasterAddressSource` callback on `Master::async_client_init_` requires a network connection request to be invoked.  The Master sends no network traffic using `async_client_init_` currently.  Only the TServer & Postgres process does.  Adding the first invocation of `Master::async_client_init_` causes the stack trace in the description.

**Comment by spolitov** [comment]
@nspiegelberg Why do you think that it does not send network traffic using it?

**Comment by spolitov** [comment]
@nspiegelberg 
`Master::Init` calls `async_client_init_->Start()`, which launched thread with `InitClient`, that performs:     
`auto result = client_builder_.Build(messenger_);`
And through `DoBuild` it invokes 
```    auto status = c->data_->SetMasterServerProxy(deadline,
            data_->skip_master_leader_resolution_,
            data_->wait_for_leader_election_on_init_);
```
Where we have `void YBClient::Data::SetMasterServerProxyAsync` calling `ReinitializeMasterAddresses`, and it has the following code block:
```  for (const auto& source : master_address_sources_) {
    auto current = source();
    full_master_server_addrs_.insert(
        full_master_server_addrs_.end(), current.begin(), current.end());
  }
```

That should use added sources.
Correct me if I'm wrong.

**Comment by bmatican** [comment]
Fixed in #4984
